File: sorting/python/sort.py
#!/usr/bin/python
import random
import time

MAX = 20000


def test(alg) -> None:
    array = [random.randint(1, 100) for _ in range(MAX)]
    print(f"Unsorted:\n{array}")
    start_time = time.time()
    array = alg(array)
    elapsed = time.time() - start_time
    print(f"Sorted:\n{array}")
    print(f"Sorting took: {elapsed * 1000000} us")


def main() -> None:
    test(merge_sort)


def merge_sort(array: list) -> list:
    n = len(array)
    if n <= 1:
        return array
    left = merge_sort(array[:n//2])
    right = merge_sort(array[(n+1)//2:])
    return merge(left, right)


def merge(left: list, right: list) -> list:
    merged = []
    n_l = len(left)
    n_r = len(right)
    n = n_r + n_l
    i, j, = 0, 0
    for k in range(n):
        if j >= n_r or (i < n_l and left[i] < right[j]):
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    return merged


if __name__ == "__main__":
    main()

File: sorting/julia/sort.jl
#!/usr/bin/julia

function merge_sort(array)
  n = size(array, 1)
  half = trunc(Int, n / 2)
  if n <= 1
    return array
  end
  left = merge_sort(array[1:half])
  right = merge_sort(array[half+1:n])
  return merge(left, right)
end

function merge(left, right)
  l, r = size(left, 1), size(right, 1)
  n = l + r
  merged = Vector{Int}(undef, n)
  i, j = 1, 1
  for k in 1:n
    if j > r || (i <= l && left[i] < right[j])
      merged[k] = left[i]
      i += 1
    else
      merged[k] = right[j]
      j += 1
    end
  end
  return merged
end

array = rand(1:100, 30)

println("Merge Sort")
println("Unsorted:")
println(array)
println("Sorted:")
println(merge_sort(array))
